signer/src/network/libp2p/utils.rs

```diff
@@ -0,0 +1,94 @@
+use std::net::ToSocketAddrs;
+
+use libp2p::multiaddr::Protocol;
+use libp2p::Multiaddr;
+use url::Url;
+
+use super::errors::SignerSwarmError;
+use super::DEFAULT_P2P_PORT;
+
+/// A trait for converting a value into a vector of [`Multiaddr`]s.
+pub trait TryIntoMultiAddrs {
+    /// Attempts to convert the value into a vector of [`Multiaddr`]s.
+    #[allow(dead_code)]
+    fn try_into_multiaddrs(&self) -> Result<Vec<Multiaddr>, SignerSwarmError>;
+}
+
+impl TryIntoMultiAddrs for Url {
+    fn try_into_multiaddrs(&self) -> Result<Vec<Multiaddr>, SignerSwarmError> {
```

Comment:
Hmmm, what are your thoughts of writing multi-addresses directly in the config? That's probably what I would do, but I'm curious about why you chose this approach?

Reply:
I chose this method because multiaddrs aren't really that popular, users will be more familiar with the standard `scheme://host:port` format in the config.

I.e., a `quic-v1` address looks like `/ip4/127.0.0.1/udp/1234/quick-v1` and tcp is `/ip4/127.0.0.1/tcp/1234` -- just easier for the user imo.

Reply:
Oh right, I also didn't want any network requirements on loading the config (i.e. dns) or for a dns resolution failure to be a config error (we can handle that gracefully at runtime).

Reply:
Hmmm, I think I understand now. There are a couple of things that seem clearer now that weren't obvious to me before:

1. This function does a DNS look-up via the `std::net::ToSocketAddrs::to_socket_addrs` function. So it requires an internet connection in order to work as designed[^1].
2. A single URL can contain a domain name, which may resolve to multiple IPv4 and/or IPv6 addresses. This implementation uses the raw IPs when constructing the multi-addresses. That's why we return a `Vec` here.
3. There is a `libp2p::core::multiaddr::from_url_lossy` function that looks like a good fit here, but it hasn't kept up with all of the supported protocols, so we need our own implementation if we want to support regular URLs.
4. A multi-address does not currently include the username, password, path, and query stuff that we can see in a URL. Multi-addresses only use the scheme, host, and port of a URL.

If we deserialize straight into a `MultiAddr` I don't think that we'd have any network requirements when loading the config. Right now I'm leaning toward expecting multi-addresses in the config as it's what we actually use. Using a URL does not allow us to use the `p2p` protocol part of a multi-address (or any other features that I do not know about). But that might be a non-issue.

Well, I get how people are more familiar with URLs, but something feels off with using them instead of multi-addresses.

[^1]: My first thought was that having the domain resolve into a static IP is an issue. The IPs associated with a domain name can change, so the IP stored locally would be incorrect. I think clients should have some expiry that requires it to re-lookup the IP associated with a domain name but if the signers have the IP fixed then that doesn't happen. This turns out to be a non-issue unless every seed address happens to have their IP changed at the same time or something silly like that.

Reply:
> If we deserialize straight into a MultiAddr I don't think that we'd have any network requirements when loading the config.

Yeah, this is right, I checked the `FromStr` code and it doesn't perform DNS lookups at this stage.

>Well, I get how people are more familiar with URLs, but something feels off with using them instead of multi-addresses.

I feel the opposite :sweat_smile: I'm generally against exposing multiaddrs in the config because:
1. They are unconventional
2. They require much more explanation in the config docs (users have to be explicit about ip4/ip6/dnsaddr)
3. We're exposing what is essentially an implementation detail of libp2p in the config (we wouldn't be using multiaddrs if we were using a `TcpListener`, for example).

> My first thought was that having the domain resolve into a static IP is an issue. The IPs associated with a domain name can change, so the IP stored locally would be incorrect. I think clients should have some expiry that requires it to re-lookup the IP associated with a domain name but if the signers have the IP fixed then that doesn't happen. This turns out to be a non-issue unless every seed address happens to have their IP changed at the same time or something silly like that. 

This is what the `identify/push` protocol you were asking about before is for, but in general yeah, unless ALL of the connected peers suddenly change their IP, then this isn't really an issue.

Reply:
Okay, I'm cool with this way.

I think we should make a few further chain to how the trait works then. Instead of returning returning multiple multi-addresses, I think we should always return one. For FQDNs in the URI we should just convert that to a FQDN multi-address.

Basically, I think we should do what [`libp2p::core::multiaddr::from_url_lossy`](https://docs.rs/libp2p/0.54.1/libp2p/multiaddr/fn.from_url.html) does, except we accept or reject different protocols. It also makes it clear what's going on, we decided to have the "address" represented as a regular [RFC 8820](https://datatracker.ietf.org/doc/html/rfc8820)[^1] URI and we map a subset of those to multi-addresses internally.

[^1]: I just took the latest RFC here, but yeah it's the regular URI that people are familiar with.

Reply:
> I think we should make a few further changes to how the trait works then. Instead of returning returning multiple multi-addresses, I think we should always return one. For FQDNs in the URI we should just convert that to a FQDN multi-address.

Yeah I can agree with this, I didn't actually know about the `dns4` and `dnsaddr` protocols when I originally wrote this -- I enabled DNS for the swarm so it should be able to resolve those itself.

And with that we can just use `Multiaddr` on the config struct and use a custom parser there (string -> url -> multiaddr), and get rid of the trait alltogether :+1:

Reply:
Alright, fixed the test issue with a "hack" in [e17719e](https://github.com/stacks-network/sbtc/pull/473/commits/e17719e79709acd2fa80ccd417e75795e67ee538).